### PR TITLE
Ground for legacy testing

### PR DESCRIPTION
**Pull request by kebeclibre**
loadViews Operational

Finally some good mapping between legacy and wowl envs

### CODE REVIEW

File: addons/web/static/tests/helpers/test_env.js [hunk 1]
```diff
@@ -17,7 +17,7 @@ odoo.define('web.test_env', async function (require) {
      * @returns {Proxy}
      */
     function makeTestEnvironment(env = {}, providedRPC = null) {
-        if (!qweb) {
+        if (!qweb && session.owlTemplates) {
```
Comment (kebeclibre) [inline addons/web/static/tests/helpers/test_env.js]: we are not sure anymore to arrive here after the loadTemplateRPC

File: addons/web/static/tests/helpers/test_utils.js [hunk 2]
```diff
@@ -135,7 +135,7 @@ odoo.define('web.test_utils', async function (require) {
 
         // this is done with the hope that tests are
         // only started all together...
-        QUnit.start();
+        //QUnit.start();
```
Comment (kebeclibre) [inline addons/web/static/tests/helpers/test_utils.js:138]: To fix, obviously

File: addons/wowl/static/src/legacy_prod_infra/legacy_setup.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
+import { Component, config, utils } from "@odoo/owl";
```
Comment (kebeclibre) [inline addons/wowl/static/src/legacy_prod_infra/legacy_setup.ts]: Split of the two files to avoid mixing production execution and testing execution

Comment (kebeclibre) [reply addons/wowl/static/src/legacy_prod_infra/legacy_setup.ts]: in particular, this file has side effects, whereas legacy.ts only exports stuff

File: addons/wowl/static/src/env.ts [hunk 4]
```diff
@@ -77,3 +77,50 @@ async function deployServices(env: OdooEnv, config: OdooConfig): Promise<void> {
     return null;
   }
 }
+
+import { OdooBrowser } from "./types";
+
+export function makeFakeLocalStorage(): OdooBrowser["localStorage"] {
```
Comment (aab-odoo) [inline addons/wowl/static/src/env.ts]: this code should be move in its own file

File: addons/wowl/static/src/env.ts [hunk 5]
```diff
@@ -77,3 +77,50 @@ async function deployServices(env: OdooEnv, config: OdooConfig): Promise<void> {
...
+    get length() {
+      return Object.keys(store).length;
+    },
+    key() {
+      return "";
+    },
+  };
+}
+
+export function makeLocalStorage(
```
Comment (aab-odoo) [inline addons/wowl/static/src/env.ts]: I don't like this function to handle both the real and mock cases. I think it should only handle the real case, and if it crashes, fallback on a RAM localStorage. On the other side, in the test environment, we should build the browser by directly using the RAM storage.

File: addons/wowl/static/tests/helpers/mock_server.ts [hunk 6]
```diff
@@ -101,3 +134,510 @@ export function makeMockServer(
...
+      if (!field) {
+        throw new Error("Field " + fieldName + " does not exist");
+      }
+      const defaultValues: any = {};
+      const stateExceptions: any = {}; // what is this ?
+      modifiersNames.forEach((attr) => {
+        stateExceptions[attr] = [];
+        defaultValues[attr] = !!field[attr];
+      });
+      // LPE: what is this ?
```
Comment (aab-odoo) [inline addons/wowl/static/tests/helpers/mock_server.ts:324]: attribute `state="open,pending"` stands for `invisible if the state isn't in [open, pending]`

File: addons/wowl/static/tests/helpers/mock_server.ts [hunk 7]
```diff
@@ -101,3 +134,510 @@ export function makeMockServer(
...
+ *
+ * @private
+ * @param {string} modelName
+ * @param {array[]} args a list with a list of fields in the first position
+ * @param {Object} [kwargs={}]
+ * @param {Object} [kwargs.context] the context to eventually read default
+ *   values
+ * @returns {Object}
+ */
+function _mockDefaultGet(
```
Comment (aab-odoo) [inline addons/wowl/static/tests/helpers/mock_server.ts:607]: this method isn't called anymore, and you know that :D

**Review by aab-odoo — COMMENTED** [review]
wow good work :-) There are details to polish of course, but this is an excellent starting point. Out of curiosity, almost all legacy tests are still commented. Is it because they don't pass, or because you simply didn't adapt them yet?

File: addons/wowl/static/src/legacy/legacy.ts [hunk 8]
```diff
@@ -55,30 +43,39 @@ odoo.define("wowl.legacySetup", async function (require: any) {
...
+      const [route, params, settings] = args;
+      wowlEnv.services.rpc(route, params, settings).then(resolve).catch(reject);
+    });
+    (prom as any).abort = rejection;
+    return prom;
+  };
+  // localStorage
+  const localStorage = wowlEnv.browser.localStorage;
+  legacyEnv.services.local_storage = Object.assign(Object.create(localStorage), {
+    getItem(key: string, defaultValue: any) {
```
Comment (kebeclibre) [inline addons/wowl/static/src/legacy/legacy.ts:76]: need a localStorage Service to:
- handle crash in Safari
- modify getItem/setItem API to JSON stringify/parse

browser should be equal to window stuff

File: addons/wowl/static/src/env.ts [hunk 9]
```diff
@@ -77,3 +77,50 @@ async function deployServices(env: OdooEnv, config: OdooConfig): Promise<void> {
...
+  localStorage?: OdooBrowser["localStorage"],
+  fake?: boolean
+): OdooBrowser["localStorage"] {
+  if (!localStorage && !fake) {
+    localStorage = window.localStorage;
+  }
+  if (localStorage && !fake) {
+    try {
+      localStorage.setItem("__localStorage__", "true");
```
Comment (kebeclibre) [inline addons/wowl/static/src/env.ts]: handle this in make env for prod

**Pull request merged**